                                                  Text Summarization Using Latent Semantic Indexing

            Table Of Contents

1. Intoduction to text summarization.

2. Importing prerequsite libraries.

3. Loading our text.

4. Text preprocessing.

5. Feature engineering.

6. Latent Sematic Indexing.

7. Leveraging Singular Value Decomposition.

8. Getting salience scores.

9. Summary of our text

                                         Text Summarization

This involves extracting the gist of a text from a document in order to lessen the length of the said document without changing the core or essence of the meaning of the text. An example is the executive summary of a research paper. We will utilize mathematics and statistical models in summarizing our document.

Text summarrization is grouped into two methodologies:

1. Extaction Based Techniques: It uses mathematical concepts like SVD(Singular Value Decomposition) to extract content such as words, key phrases and sentences that retains thr original meaning of the document thus creating a short executive summary of the text.

2. Abstraction Based Techiniques: This involves complex algorithms learning the language semantics of a text in order to generate a representative summary of the original text without compromising the original meaning of the text.

We will implement our text summarization using extraction based technique where we will leverage on SVD to summarize our document.

In [ ]:
# importing necessary packages
import nltk 
import numpy as np 
import re
import pandas as pd

We will use an excerpt extracted from Blackwell-Companion-to-the-Romanian Empire in subtopic: The Orgaization of Estates, page 326 - 331.

In [53]:
data ="""
The strategies that Roman landowners followed in managing their wealth had to be
adapted both to the geographical conditions affecting agricultural production and the
overall climate for investment in agriculture in the Roman Empire. Perhaps the
greatest boom in agriculture, in Italy at least, occurred during the last two centuries
of the republic and the first century CE, when an increasingly wealthy upper class
invested its wealth in land and was able to make great profits by supplying agricultural
products to the growing urban population in Rome and other Italian cities. These
landowners developed the so-called villa system of agriculture, characterized by
intensively cultivated estates that produced wine and other cash crops. This type of
agriculture was associated especially with coastal regions of Italy, which provided easy
communications by sea with urban markets (Morley 1996: 129–38). The labor force
on these estates consisted primarily of slaves, under the supervision of a bailiff, or
vilicus, who was often also a slave. These estates were compact in size, comprising at
most several hundred hectares. The wealthiest landowners might possess numerous
compact estates, often located in diverse regions within Italy. This pattern of ownership lessened the landowners’ exposure to the substantial risk posed by the irregularity of rainfall in Roman agriculture. Estates like the Villa Settefinestre at Cosa in
Etruria, the most carefully excavated Roman estate, typically included a residence, the
pars urbana, that served to accommodate the owner during visits to the estate
(Carandini 1989). These villae became increasingly luxurious over the course of the
late republic and the first century, and the lines between the economic and social
function of an estate could become blurred. The pars rustica of such an estate would
commonly include pressing facilities and buildings that could accommodate a substantial labor force of slaves.
300 Dennis P. Kehoe
The landowner’s profit from this type of estate depended on concentrating the
efforts of the slave staff on the cultivation of cash crops, especially wine and olive oil.
At the same time, the slaves could be kept fully occupied throughout the agricultural
year by producing most if not all of the food that the estate consumed; in this way the
estate remained largely autonomous. Because of the seasonal nature of agricultural
labor, it was not economical to maintain a slave staff that could perform all of the labor
on the estate throughout the year. Instead of carrying the expense of keeping slaves
who would not be fully employed, landowners would recruit additional labor at the
harvest and other busy seasons from nearby small landowners or even tenants. Thus
in coastal Italy, at least, an area with probably the largest concentration of estates
owned by the upper classes, small-scale peasant agriculture remained a basic feature of
the economy throughout the early imperial period (Rathbone 1981).
Slaves were often maintained under conditions that were extraordinarily cruel by
modern standards. Many agricultural slaves, called vincti, were kept in chains and
housed in slave-barracks, or rather jails, called ergastula. These were dimly lit buildings that provided slaves with basic shelter and few comforts, if any. At the same time,
slaves were subject to violent punishment if they shirked their labor (Bradley 1994).
Because the profit of the villa depended on maximizing the effort extracted from the
slaves, estate owners were concerned to keep their slaves busy throughout the year.
The Roman agriculture writer Cato, a senator writing in the second century BCE,
exemplifies this attitude in his prescriptions for the estate owner’s correct management of his slave labor force. In Cato’s view, when rainy weather or storms made it
impossible to undertake normal agricultural duties, the slaves were to be kept busy by
working indoors on such tasks as cleaning storage jars, sweeping out the villa, moving
grain, carrying dung out to the dung heap, cleaning the seed, repairing ropes, and
mending clothing. Even on holidays, the slaves were to be kept busy by cleaning
ditches, repairing roads, and cleaning up the gardens and other farmland. When the
slaves were sick, they were not to be given their normal food ration. The point is
that slaves represented a commodity, one that entailed a fixed cost to the landowner
(Agr. 2.3–5). Some slaves kept under these conditions might be worked to death.
In the empire, it seems likely that the exploitation of slaves was modified somewhat
as they became more expensive. The end of the republican wars of conquest reduced
the importation of large numbers of slaves into Italy. As a consequence, agricultural
slaves were increasingly bred on Roman estates. However, it remains a hotly debated
topic to what extent slave breeding met the Roman Empire’s demand for slaves.
Some of the demand for slaves was met by imports from outside the empire. Within
the empire, exposed children might be raised as slaves, and others might be sold into
slavery (Scheidel 1997; cf. W. V. Harris 1999). The agronomist Columella, writing in
the first century CE, offered special incentives to female slaves who produced offspring (Rust. 1.8.19); Columella may have been imitating the emperor Augustus’
marriage legislation, which offered privileges to women who produced three children. But we should not exaggerate the improvement in the situation of slaves. Pliny
the Younger, as we will see, did not maintain chained slaves, but he was aware that this
practice was still common in Italy (Ep. 3.19.7).
In his recommendations about the proper treatment of slaves, Columella indicates
that the relationship between landowner and slave was in essence one of exploitation,
no matter how kind the landowner might seem. Columella exhibits a seemingly more
Landlords and Tenants 301
humane attitude towards slaves than Cato, but at the same time his concern is still to
maximize the effort that could be obtained from them:
As regards other slaves these precepts are generally to be preserved, and I do not regret
maintaining them, namely, that I speak on rather familiar terms more often to the
country slaves than to the household slaves, as long as they have not comported
themselves badly, and, since I understand that because of this friendliness on the owner’s
part their perpetual labor is lightened, I might sometimes joke with them and permit
them to joke more. I now often follow this practice, that I deliberate about some new
tasks with them as if they were more experienced, and through this I learn the nature of
each one’s mind and how intelligent it is. Then I see them more willingly approach a task
that they think they have been consulted about and that they think has been undertaken
on their recommendation. (Rust. 1.8.15)
Columella’s seeming friendliness towards his slaves was all part of a general policy to
manage slaves with a system of rewards and punishments. He still kept slaves in
chains, and fair treatment of slaves was recommended only in order to avoid making
his labor force intractable through the excessive cruelty of their overseers (Rust.
1.8.16–19).
Since the villa system generated enormous wealth for some landowners, it is
noteworthy that it was not exported to the provinces in the early empire. In fact,
the villa system in Italy declined rather precipitously beginning in the late first century
CE. To be sure, the success of the classical villa system seems to have been the product
of a particular set of circumstances: an increasingly wealthy group of landowners,
rapid expansion of the urban population in Rome, and the creation of new markets,
most notably in Gaul. These particularly favorable circumstances could not endure
indefinitely. In the late first and second centuries, the villa system underwent a
substantial transformation, with the compact villas in many locations giving way to
larger estates cultivated less intensively. This transformation has been interpreted as a
sign of a major crisis in Italian agriculture, for which several explanations have been
offered. One explanation is that it was no longer profitable for landowners to
maintain large numbers of slaves in the face of competition from provincial landowners, especially in Gaul and Spain. Indeed, imports of wine into Rome from Gaul
and Spain increased in the early empire, while Spain also became one of the empire’s
leading suppliers of olive oil. At the same time, we should not overstate the nature of
the crisis, especially the extent to which the transformation of the villa system
impinged on the fortunes of upper-class landowners (Morley 1996: 135–42). In
the early empire, the rapid growth of Rome as a consumer city ended, while the
class of elite landowners spread outward from Italy to the western and then to the
eastern provinces.
The emerging class of elite provincial landowners derived its income by supplying
what was by the second century CE a more stable but still considerable urban
market for the same agricultural products that their Italian counterparts had
produced through the villa system in the late republic. Often it was small-scale tenants
cultivating farms within larger estates who produced these cash crops. The estates of
upper-class landowners in the Roman Empire varied considerably from region
to region in their composition and organization. Italy in the early imperial period
is famous for its extensive estates, or latifundia. In fact, estates that we might
302 Dennis P. Kehoe
characterize as latifundia consisted of numerous much smaller farms, or fundi, that
might not even be contiguous. Such estates grew over time as their owners purchased
or inherited individual small farms. This is the pattern of landownership revealed
by the famous alimentary table from Veleia in northern Italy, which records the
properties pledged by landowners in the town to support a program to provide
food for children (De Neeve 1984: 224–30).
The case of Pliny the Younger sheds light on the patterns of upper-class land
ownership in Italy. Pliny derived his income from estates in two locations, his home
town of Comum in northern Italy, and the Umbrian town of Tifernum Tiberinum.
At the latter location, Pliny’s estate consisted of a number of individual farms that he
describes as scattered among parcels belonging to an adjacent estate whose purchase
he was considering (Ep. 3.19.1; De Neeve 1990: 373). These estates were under the
management of employees called actores, who were probably either slaves or freedmen of Pliny. These actores, in turn, reported to Pliny’s business managers, or
procurators. A slave-labor force under the supervision of the vilici may have worked
some of the land on Pliny’s estate, but the bulk of Pliny’s land was leased out to
tenants, who cultivated the individual fundi.
The juxtaposition of large estates and a small-scale peasant agricultural economy is
even more likely to have characterized other regions of Italy and the provinces.
During the early imperial period the western provinces, especially Spain, southern
Gaul, and Africa provided fertile recruiting grounds for the Roman elite, contributing
large numbers of families to both the senatorial and equestrian orders. This development was surely a product of the development of the agricultural economy in these
provinces, with a class of large landowners exercising an increasing dominance in the
rural economy. Early imperial Gaul saw a distinct transformation in the countryside,
as numerous dispersed villas gradually supplanted nucleated hilltop settlements.
These villas were decorated on a scale not seen before in Gaul, and they indicate
that the wealthier landowners in Gaul were adopting a basic Roman cultural form.
The Gallic villas, however, were often modest in size, in some regions ranging from
50 to 100 hectares (Woolf 1998: 142–68). The evidence for these villas is largely
archaeological, and we have very little information as to how they were cultivated,
whether by slaves, tenants, day-laborers, or some combination of all three. Recent
archaeological work in Africa confirms the impression offered by inscriptions and
literary evidence that this region of the Roman Empire also saw the growth of
substantial estates, often devoted to the cultivation of olives as a cash crop
(D. J. Mattingly 1988c; M. De Vos 2001: 21–9).
Some parts of the empire saw the development of large private estates comparatively late, as in Egypt, where such estates began to appear in the second and third
centuries (Rathbone 1991), or Palestine, where large estates apparently only formed
in the third and fourth centuries (Z. Safrai 1994: 322–64). In these areas, the absence
of large landowners could mean that other institutions played a significant role in
setting the conditions under which agriculture was practiced. In Egypt, agricultural
land in the Nile valley and Fayum was often divided into relatively small parcels,
separated by embankments that were designed to capture the Nile flood. Grain was
the usual crop in such land. Land not reached by the Nile flood needed to be irrigated
artificially. Often irrigated land was devoted to orchards and vineyards. Much of the
land was privately owned, but the state also maintained control over a great deal of
Landlords and Tenants 303
land in most if not all nomes (administrative divisions). This circumstance, in all
likelihood, promoted greater equality of land distribution, since state land in Egypt
was customarily leased out in small plots to individual cultivators, who enjoyed
security of tenure and could often pass their cultivation rights on to their heirs
(Rowlandson 1996: 70–101). Egyptian agriculture required the continued maintenance of irrigation dikes and drainage canals to make sure that the flood reached all
the land that had to be irrigated and also that the water could be drained to allow the
sowing of the crop. This work was organized communally at the village level, and
under Roman rule the state imposed the task of maintaining the irrigation facilities on
the farming population as a liturgy. Thus most male peasants in Roman Egypt were
required to provide five days of labor each year working on the dikes and canals
(Lewis 1983: 107–33). In the more intensive agriculture of Egypt, wheat yields were
substantially higher than elsewhere in the empire, on the order of perhaps 1,000 kg/
ha (Rathbone 1991: 242–4, 465).
On occasion, documentary papyri make it possible to examine the pattern of
landownership to a degree of detail not possible elsewhere in the Roman Empire.
In an analysis of landholding in the Hermopolite nome in the fourth century, for
example, Roger Bagnall concludes that the majority of landowners owned plots of
land between ten and 50 arouras in size (10 arouras is approximately 2.5 hectares, an
amount of land probably sufficient to support an individual family). Only a few
villagers might own land in excess of this amount, but the wealthiest landowners in
the village were probably residents of the nome metropoleis. These landowners did not
own land in the village on a larger scale than the largest landowners among the
villagers, but in all likelihood such landowners had holdings in many villages at the
same time, so that their level of wealth probably exceeded that of even the very
wealthiest villagers (Bagnall 1992). This pattern of landownership seems to have been
approximated in other areas in Egypt. One location that has been subject to recent
study is the city and surrounding territory of Oxyrhynchus (Rowlandson 1996:
102–38). The Oxyrhynchite nome had a population on the order of 100,000–
120,000. There too, the vast majority of landowners had relatively modest holdings
of 5–50 arouras, while a much smaller number of landowners had 500 arouras or
more of land (c.125 hectares).
The village also provided the organizational framework for larger estates in Roman
Egypt. The private estates that developed typically consisted of a variety of individual
parcels scattered among other lands in a village. There is little evidence for the use of
slaves in agriculture in Egypt, certainly not in the way in which slaves were employed
in Italy. Often tenancy provided the means for the owners of scattered parcels, like the
urban landowners at Oxyrhynchus, to generate an income from such lands. But as the
largest landowners gained control of increasingly large holdings, they began to realize
some economies of scale. In the second and third centuries, some estates had a central
facility, or epoikion, that housed workers on the estate and might also include facilities
for pressing grapes and olives (cf. Banaji 2001: 174–6 for epoikia in late antique
Egypt). The best attested of these estates is the one belonging to Aurelius Appianus, a
third-century landowning magnate from Alexandria. The organization of this estate is
attested in great detail by the Heroninos archive, a collection of some 450 documents, including drafts of monthly accounts and correspondence associated with the
working of the estate. The Heroninos archive takes its name from the manager of
304 Dennis P. Kehoe
the division, or phrontis, of the estate located at the village of Theadelphia (Rathbone
1991). These divisions consisted of numerous individual parcels cultivated with
wheat, fodder crops, and vines. At the level of the village divisions, the labor on the
individual parcels was supplied by a small number of permanent workers, supplemented by numerous workers hired from the village on a daily basis. Presumably these
workers were recruited from smaller landowners and tenants working in the village, as
well as from the landless rural population. Some parcels whose cultivation could not
be integrated into the village division were leased out to tenants. The central management of the estate, to which the individual managers reported, maintained control
over key resources, such as draft animals, and it supervised the workers who managed
these resources. This system of centralized management made it possible to keep the
costs of production in the individual divisions under control by sharing centrallymaintained resources that the individual divisions would otherwise have to duplicate """
# converting our text into a bunch of sentences
sentences = nltk.sent_tokenize(data)
# printing the length of our sentences
print(len(sentences))

.
116


Our text has a total of 116 sentences. Let us print it in a more readable format.

In [54]:
#checking our text
print(data)


The strategies that Roman landowners followed in managing their wealth had to be
adapted both to the geographical conditions affecting agricultural production and the
overall climate for investment in agriculture in the Roman Empire. Perhaps the
greatest boom in agriculture, in Italy at least, occurred during the last two centuries
of the republic and the first century CE, when an increasingly wealthy upper class
invested its wealth in land and was able to make great profits by supplying agricultural
products to the growing urban population in Rome and other Italian cities. These
landowners developed the so-called villa system of agriculture, characterized by
intensively cultivated estates that produced wine and other cash crops. This type of
agriculture was associated especially with coastal regions of Italy, which provided easy
communications by sea with urban markets (Morley 1996: 129–38). The labor force
on these estates consisted primarily of slaves, under the supervision of a ba

               Text Processing

Before we implement document summarrization, we will perform  text wrangling which will involve:

1. Removing special characters and whitespaces in our text.
2. Changing our text from proper case to lower case.
3. Removing empty long spaces between words.
4. Reducing sentences into words and removing words that have insignificant meaning (stopwords) like articles, conjuctions etc and retaining those that are important in our text.
5. Joining the cleaned words into sentences.


In [55]:
stopwords = nltk.corpus.stopwords.words('english')
# a function to clenase our text
def normalize_doc(doc):
    # removing whitespaces and special charactes
    doc = re.sub(r'[^a-zA-Z\s]', "", doc, re.I|re.A)
    # chaing to lower case
    doc = doc.lower()
    doc = doc.strip()
    # tkenizing text
    tokens = nltk.word_tokenize(doc)
    #removing stopwords
    filtered_words = [token for token in tokens if token not in stopwords]
    # joining words
    doc = " ".join(filtered_words)
    return doc
#
normalize_corpus = np.vectorize(normalize_doc)

In [56]:
# Appling our function to cleanse our text
sentences = nltk.sent_tokenize(data)
norm_sentence = normalize_corpus(sentences)

            Feature Engineering

This is a process of extracting meaningful attributes from raw textual data and feed it into a statitical algorithm that transforms each term/word into a vector. This is done since machine learning models cannot summarize on raw text.

We will use TFIDVectorizer from sklearn which performs the following functions:


1. Terms Frequency. This involves is the frequency of a word/term in a document.
2. Inverse document frequency. This involves computing by dividing the total number of documents in our corpus by the document frequency for each term and then applying logarithmic scaling to the result.

In [57]:
from sklearn.feature_extraction.text import TfidfVectorizer
# initializing TFIDVectorizer
# using normalizing our results using l2 norm
tv = TfidfVectorizer(min_df=0,max_df=1, norm='l2')
dt_matrix = tv.fit_transform(norm_sentence)
# creating a feature matrix which maps a collection of documents into features (row = document, column= feature)
dt_matrix = dt_matrix.toarray()
# getting various vector names in our text
vocabs = tv.get_feature_names_out()
# transposing our matrix
td_matrix = dt_matrix.T
print(td_matrix.shape)
# showing a feature engineered dataframe
pd.DataFrame(np.round(td_matrix, 2), index=vocabs).head(10)

(813, 116)


0    1    2    3    4    5    6    7     8    9    ...  106  \
able            0.00  0.2  0.0  0.0  0.0  0.0  0.0  0.0  0.00  0.0  ...  0.0   
absence         0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.00  0.0  ...  0.0   
accommodate     0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.22  0.0  ...  0.0   
accounts        0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.00  0.0  ...  0.0   
actores         0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.00  0.0  ...  0.0   
adapted         0.27  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.00  0.0  ...  0.0   
additional      0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.00  0.0  ...  0.0   
adjacent        0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.00  0.0  ...  0.0   
administrative  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.00  0.0  ...  0.0   
adopting        0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.00  0.0  ...  0.0   

                107   108  109  110  111  112  113  114  115  
able            0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
absence         0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
accommodate     0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
accounts        0.0  0.26  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
actores         0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
adapted         0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
additional      0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
adjacent        0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
administrative  0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
adopting        0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[10 rows x 116 columns]

          Latent Semantic Indexing

This method utilizes the principle that a hidden structure exists among terms that are related contextually and hence should also be correlated in the same singular space. It utilises SVD(Singular value decomposition) which is a linear algebra technique of factorizing a matrix e.g matrix M. I wont delve into it but, ""Mathematically, the matrix M can be represented using SVD as 
a factorization such that:
Mm*n = Um*m Sm*n V^t n*n


Where we have the following decompositions:
1. U is a m × m unitary matrix such that UTU = Im × m where I is the 
identity matrix. The columns of U indicate left singular vectors.
2. S is a diagonal m × n matrix with positive real numbers on the 
diagonal of the matrix. This is also often also represented as a vector 
of m values which indicate the singular values.
3. VT is a n × n unitary matrix such that VTV = In × n, where I is the identity 
matrix. The rows of V indicate right singular vectors.
This tells us that U and V are orthogonal. The singular values of S are particularly 
important in summarization algorithms. We use SVD particularly for low rank matrix 
approximation, where we approximate the original matrix M with a matrix Mˆ such that 
this new matrix is a truncated version of the original matrix M with a rank k and can 
be represented by SVD as M U ˆ ˆ
SVT = , where ˆ
S is a truncated version of the original S
matrix and now consists of only the top k largest singular values and the other singular 
values are represented by zero""

In [58]:
from scipy.sparse.linalg import svds
# creating a unction that will leverage SVD
def low_rank_svd(matrix, singular_count):
    u, s, vt = svds(matrix, k = singular_count)
    return u,s,vt

In [67]:
# specifying the number of sentences to be gotten from the text
num_sentence = 12
num_topics= 3
# initializing keys to our function
u, s,vt = low_rank_svd(td_matrix, singular_count=num_topics)
print(u.shape, s.shape, vt.shape)
#
term_topic_mat, singular_values, topic_document_mat = u, s, vt 

(813, 3) (3,) (3, 116)


We will apply a threshold-based approach to remove singular values that are less than 80% of the largest singular value

In [69]:
# specifying the minimum threshold
sv_threshold = 0.8
# applyng our threshold to our ingular values
min_sigma_values = max(singular_values) * sv_threshold
singular_values[singular_values < min_sigma_values] = 0

          Salience Scores

This will involve calculating salience scores for each sentence, sort them by descending order and pick sentences that correspond to the highest scores in order to make a summary of our textual data/document.

Getting salience scores involves computing the sum of sentence weights of terms(multiplication of singular values and sentence vectors from V^t) and take the squareroot.

In [70]:
# claculating salience scores
saliencyscores = np.sqrt(np.dot(np.square(singular_values), np.square(topic_document_mat)))
saliencyscores

array([2.08960310e-01, 2.32296005e-01, 3.40987348e-01, 1.63674015e-01,
       3.26040294e-01, 1.68042177e-01, 3.45197107e-01, 1.58699053e-01,
       1.78874945e-01, 1.50311351e-01, 3.23999771e-01, 2.26894885e-01,
       3.84955029e-01, 3.18432569e-01, 2.41057069e-01, 2.41432536e-01,
       1.97183687e-01, 2.51997420e-01, 1.83797302e-01, 2.57927071e-01,
       3.94085937e-01, 2.68948948e-01, 1.94106542e-01, 2.28532215e-01,
       2.21708142e-01, 1.81071790e-01, 5.83826077e-16, 3.24649767e-01,
       2.14508631e-01, 1.92542130e-01, 3.32475329e-01, 2.35868088e-01,
       2.72297364e-01, 2.08974662e-01, 1.23359821e-17, 2.22510767e-01,
       6.42281496e-02, 2.29838653e-01, 2.13956946e-01, 3.80796608e-17,
       2.27552842e-01, 3.67575499e-01, 5.63289030e-02, 1.00540669e-02,
       9.78680128e-02, 2.77247109e-01, 3.83002108e-01, 1.58687795e-16,
       2.24242422e-01, 1.91976374e-01, 1.76075146e-01, 4.47203417e-02,
       2.74645197e-01, 8.42418587e-02, 2.46818237e-01, 1.57356227e-01,
      

In [71]:
# sorting top 12 sentences with highest scores.
top_12_indices = (-saliencyscores).argsort()[:num_sentence]
top_12_indices.sort()
# joining them to form a summary
print("\n".join(np.array(sentences)[top_10_indices]))
#print(np.array(sentences)[top_10_indices])

These
landowners developed the so-called villa system of agriculture, characterized by
intensively cultivated estates that produced wine and other cash crops.
The wealthiest landowners might possess numerous
compact estates, often located in diverse regions within Italy.
At the same time, the slaves could be kept fully occupied throughout the agricultural
year by producing most if not all of the food that the estate consumed; in this way the
estate remained largely autonomous.
Because the profit of the villa depended on maximizing the effort extracted from the
slaves, estate owners were concerned to keep their slaves busy throughout the year.
As a consequence, agricultural
slaves were increasingly bred on Roman estates.
Columella exhibits a seemingly more
Landlords and Tenants 301
humane attitude towards slaves than Cato, but at the same time his concern is still to
maximize the effort that could be obtained from them:
As regards other slaves these precepts are generally to be preserve

Wow! and there we have a summary of our text.

References:
1. https://www.google/Dipanjan-Sarkar-Text-Analytics-with-Python.
2.